From 3228547f038aab9dca80f2ebc2b0b7dc2d9818ae Mon Sep 17 00:00:00 2001
From: morganamilo <morganamilo@archlinux.org>
Date: Sun, 29 Jan 2023 14:04:34 +0000
Subject: [PATCH] makepkg: add -D flag to change directory before building

This is similar to -C in git/make/nina. Sadly -C is already taken for
us.

This is useful for scripts where you for loop over packages, as well as
when I'm testing makepkg builds and I'm too lazy to cd.
---
 doc/makepkg.8.asciidoc |  4 ++++
 scripts/makepkg.sh.in  | 13 ++++++++++---
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/doc/makepkg.8.asciidoc b/doc/makepkg.8.asciidoc
index f350cc37..014c07d1 100644
--- a/doc/makepkg.8.asciidoc
+++ b/doc/makepkg.8.asciidoc
@@ -146,6 +146,10 @@ Options
 *-C, \--cleanbuild*::
 	Remove the $srcdir before building the package.
 
+*-D* <dir>, *\--dir* <dir> ::
+	Cd into <dir> before building the package. Files that would be created in
+	the currently directory will instead be created in the specified directory.
+
 *\--allsource*::
 	Do not actually build the package, but build a source-only tarball that
 	includes all sources, including those that are normally downloaded via
diff --git a/scripts/makepkg.sh.in b/scripts/makepkg.sh.in
index 2e5a90f2..37437b78 100644
--- a/scripts/makepkg.sh.in
+++ b/scripts/makepkg.sh.in
@@ -43,7 +43,6 @@ unset GREP_OPTIONS
 declare -r makepkg_version='@PACKAGE_VERSION@'
 declare -r confdir='@sysconfdir@'
 declare -r BUILDSCRIPT='@BUILDSCRIPT@'
-declare -r startdir="$(pwd -P)"
 
 LIBRARY=${LIBRARY:-'@libmakepkgdir@'}
 
@@ -51,6 +50,7 @@ LIBRARY=${LIBRARY:-'@libmakepkgdir@'}
 ASDEPS=0
 BUILDFUNC=0
 BUILDPKG=1
+CHDIR=''
 CHECKFUNC=0
 CLEANBUILD=0
 CLEANUP=0
@@ -1049,8 +1049,8 @@ fi
 ARGLIST=("$@")
 
 # Parse Command Line Options.
-OPT_SHORT="AcCdefFghiLmop:rRsSV"
-OPT_LONG=('allsource' 'check' 'clean' 'cleanbuild' 'config:' 'force' 'geninteg'
+OPT_SHORT="AcCdD:efFghiLmop:rRsSV"
+OPT_LONG=('allsource' 'check' 'clean' 'cleanbuild' 'config:' 'dir:' 'force' 'geninteg'
           'help' 'holdver' 'ignorearch' 'install' 'key:' 'log' 'noarchive' 'nobuild'
           'nocolor' 'nocheck' 'nodeps' 'noextract' 'noprepare' 'nosign' 'packagelist'
           'printsrcinfo' 'repackage' 'rmdeps' 'sign' 'skipchecksums' 'skipinteg'
@@ -1081,6 +1081,7 @@ while true; do
 		--check)          RUN_CHECK='y' ;;
 		--config)         shift; MAKEPKG_CONF=$1 ;;
 		-d|--nodeps)      NODEPS=1 ;;
+		-D|--dir)         shift; CHDIR=$1 ;;
 		-e|--noextract)   NOEXTRACT=1 ;;
 		-f|--force)       FORCE=1 ;;
 		-F)               INFAKEROOT=1 ;;
@@ -1127,6 +1128,12 @@ while [[ $1 ]]; do
 	shift
 done
 
+if [[ -n $CHDIR ]]; then
+	cd_safe "$CHDIR"
+fi
+
+declare -r startdir="$(pwd -P)"
+
 # setup signal traps
 trap 'clean_up' 0
 for signal in TERM HUP QUIT; do
-- 
2.39.1.418.g7876265d61

